File: leetcode_2022/26.minimumCostOfClimbing.js
//for a given staircase, the i-th step is assigned a non-negative cost indicated
//by a cost array.
//once you pay the cost for a step, ou can either climb one or two steps.
//Find the minimum cost to reach the top of the staircase. 
//Your first step can either be the fist or second step.

const cost = [20,15,30,5]

const minCostClimbingStairs = (cost) => {
  const n = cost.length;
  const dp = [];
  return Math.min(minCost(n-1, cost, dp), minCost(n-2, cost, dp))
}

const minCostClimbingStairs2 = (cost) => {
  const n = cost.length;
  const dp = [];
  for(let i = 0; i < n; i++){
    if(i < 2){
      dp[i] = cost[i];
    } else {
      dp[i] = cost[i] + Math.min(dp[i-1], dp[i-2])
    }
  }
  return Math.min(dp[n-1], dp[n-2]);
}

console.log(minCostClimbingStairs2(cost))

const minCost = (i, cost, dp) => {
  if (i < 0 )return 0;
  if (i === 0 || i === 1) return cost[i];
  if (dp[i] != undefined) return dp[i];
  
  dp[i] = cost[i] + Math.min(minCost(i-1, cost, dp), minCost(i-2, cost, dp))
  return dp[i]
}

console.log(minCostClimbingStairs(cost))

File: leetcode_2022/26.minimumCostOfClimbing.py
cost = [20,15,30,5]

def minCostClimbingStairs (cost):
    
    def find_cost (i, memo = {}):
        if i in memo:
            return memo[i]
        if i <= 1:
            return cost[i]
        memo[i] = cost[i] + min(find_cost(i - 1, memo), find_cost(i - 2, memo))
        return memo[i]
    
    length = len(cost)
    return min(find_cost(length - 1), find_cost(length - 2))

    
print(minCostClimbingStairs(cost))

def minCostClimbingStairs2 (cost):
    dpOne = cost[0]
    dpTwo = cost[1]
    
    for i in range(2, len(cost)):
        current = cost[i] + min(dpOne, dpTwo)
        dpOne = dpTwo
        dpTwo = current
    return min(dpOne, dpTwo)

print(minCostClimbingStairs2(cost))